### PR TITLE
[Test] Cypress Exam Participation

### PR DESCRIPTION
**Pull request by jhainzlschmid**
### Motivation and Context

We want to have an End-to-End test, where a student participates in an exam.

### Description

This PR adds a new test, which creates a running exam with a text and programming exercise, registers a student and participates as said student in the exam. Pageobjects have been added to aid with the navigation in the exam as well als the course overview in the beginning. Further exercise types like quizzes will be added later once we have the page objects for those.

### Steps for Testing

Tests are executed by [bamboo](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-208), so only code reviews are need

### Review Progress

- Code Review
  - [x] Review 1
  - [ ] Review 2

### CODE REVIEW

**Comment by jhainzlschmid** [comment]
Waiting for [PR #3908](https://github.com/ls1intum/Artemis/pull/3908) to be merged

**Comment by MaisiKoleni** [comment]
@jhainzlschmid @rs-maker Could you work out how the PRs should proceed? Especially because this one is blocked by #3908, but #3908 is not in the milestone this one is in. To unblock the milestone, either this needs to move out of milestone [5.1.4](https://github.com/ls1intum/Artemis/milestone/126) or the other PR needs to be part of it as well.

**Comment by jhainzlschmid** [comment]
[Tests still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-244) after merging develop

File: src/test/cypress/fixtures/requests/programming_exercise_exam_template.json
Comment (rs-maker) [inline src/test/cypress/fixtures/requests/programming_exercise_exam_template.json:21]: Why is this template necessary? we already have a programming_exercise_template

File: src/test/cypress/integration/exam/ExamDateVerification.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamDateVerification.spec.ts:95]: <a href="https://github.com/ls1intum/Artemis/pull/3918#discussion_r706851196"> Here </a> you asked me to create a textfile or a json file which would contain our text exercise submissions. I think it would also be prudent to do this here instead of having this long string in the code.

File: src/test/cypress/integration/exam/ExamDateVerification.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamDateVerification.spec.ts:129]: Same as above.

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamParticipation.spec.ts:24]: Can be refactored with default values.

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamParticipation.spec.ts:25]: I would prefer for the submission text to come from a fixture.

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamParticipation.spec.ts:50]: Is there no problem if all exercise groups have the same name? Honest question this just surprises me.

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (rs-maker) [inline src/test/cypress/integration/exam/ExamParticipation.spec.ts:107]: I think im guilty of doing this too but this assertion is essentially redundant because the test fails on a non 2xx response code anyways. Unless you wanted to check that the code is exactly 200.

File: src/test/cypress/support/pageobjects/exam/ExamStartEndPage.ts
Comment (rs-maker) [inline src/test/cypress/support/pageobjects/exam/ExamStartEndPage.ts:18]:
```suggestion
        cy.intercept(POST, BASE_API + 'courses/*/exams/*/student-exams/submit').as('finishExam');
```

File: src/test/cypress/support/pageobjects/exercises/text/TextEditorPage.ts
Comment (rs-maker) [inline src/test/cypress/support/pageobjects/exercises/text/TextEditorPage.ts:14]:
```suggestion
        cy.intercept(PUT, BASE_API + `exercises/*/text-submissions`).as('savedSubmission');
```

File: src/test/cypress/support/pageobjects/exercises/text/TextEditorPage.ts
Comment (rs-maker) [inline src/test/cypress/support/pageobjects/exercises/text/TextEditorPage.ts:15]: Better selector?

**Review by rs-maker — CHANGES_REQUESTED** [review]
Left some comments ;)

File: src/test/cypress/integration/exam/ExamDateVerification.spec.ts
Comment (jhainzlschmid) [reply src/test/cypress/integration/exam/ExamDateVerification.spec.ts:95]: Sure makes sense, but that shouldn't be part of this PR since that code section wasn't added by my PR. I will add it in the [text exercise assessment PR](https://github.com/ls1intum/Artemis/pull/3963)

File: src/test/cypress/integration/exam/ExamDateVerification.spec.ts
Comment (jhainzlschmid) [reply src/test/cypress/integration/exam/ExamDateVerification.spec.ts:129]: Same here

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (jhainzlschmid) [reply src/test/cypress/integration/exam/ExamParticipation.spec.ts:50]: The default value generates a uuid, so it's not always the same name

File: src/test/cypress/integration/exam/ExamParticipation.spec.ts
Comment (jhainzlschmid) [reply src/test/cypress/integration/exam/ExamParticipation.spec.ts:107]: I think in this case cypress does not apply the 2xx or 3xx validation because it is not a request coming from cypress. Intercepted requests don't have that validation as far as I know

File: src/test/cypress/fixtures/requests/programming_exercise_exam_template.json
Comment (jhainzlschmid) [reply src/test/cypress/fixtures/requests/programming_exercise_exam_template.json]: Good catch. I think it was a relic, which is not needed anymore (It's not referenced in any code

**Comment by jhainzlschmid** [comment]
[Test still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-254) after implementing suggested changes

File: src/test/cypress/integration/exam/ExamDateVerification.spec.ts
Comment (jhainzlschmid) [reply src/test/cypress/integration/exam/ExamDateVerification.spec.ts]: I added this in this PR nevertheless because another spec file needed the fixture as well

**Review by rs-maker — APPROVED** [review]
Looks good.

**Comment by jhainzlschmid** [comment]
[Test still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-255) after resolving the merge conflict

**Review by rs-maker — APPROVED** [review]
Re-approve.

**Pull request merged**
